### DISCUSSION TITLE
GitHubのAPIマニュアルからカチャカへ接続

### DISCUSSION THREAD

**Discussion by hhs0172282066** [pf-robotics/kachaka-api#60] [General]
GitHubのAPIマニュアルからカチャカへ接続して、
JupyterLabからサンプルプログラムの実行はできました。
client.get_shelves()
client.get_locations()
client.move_shelf("S01","L01")
client.return_shelf("S01")　
など。
これらを外部からの信号（ネットワーク経由）で動かしたいのですが、
なにか良いソフト・手順等があれば教えていただきたいです。

**Comment by nozaki-pfr** [comment]
カチャカを利用いただきありがとうございます。
ケースごとにサンプル例を紹介致します。

## 外部からカチャカのIPアドレスでカチャカに到達できる場合

### gRPCを外部から叩く

#### kachaka-apiライブラリ

例えば外部でpythonプログラムやJuptyerLabを動かして
```
client = kachaka_api.KachakaApiClient("<カチャカのIPアドレス >:26400")
```
とすることでかカチャカの中で動かすのと同様に通信を行うことができます。

#### gRPC-web

gRPC-webを利用してwebからkachaka-apiを利用するsampleが下記にあります
https://github.com/pf-robotics/kachaka-api/tree/main/web/demos/kachaka_api_web_sample

### カチャカにサーバを立ててアクセスする

URLからkachaka-apiを実行する簡単なサンプルが下記にあります。
https://github.com/pf-robotics/kachaka-api/blob/main/python/demos/url_kachaka_api.ipynb


## 到達できない場合

mqttという通信方式を利用して、mqttのブローカー経由で司令を出すことができます。
下記にmqttを利用してスマートスピーカーからカチャカに司令を送るサンプルがあります。
https://github.com/pf-robotics/kachaka-api/blob/main/python/demos/README_SMART_SPEAKER.md

**Reply by hhs0172282066** [reply]
ネットワーク上からカチャカにアクセス可能な状態で、
JuptyerLabから

client = kachaka_api.KachakaApiClient("<カチャカのIPアドレス >:26400)

を実行するとエラーになります。
※上記コードは回答に記述されているものをコピーしたもの
※Jupyterlabバージョン 4.0.3

”を削除したり〈〉の有無でもシンタックスエラーが発生します。
引数部分に入力があると受け付けないようですが、
何か対処方法はありますでしょうか？

**Reply by nozaki-pfr** [reply]
すみません、後半の"が抜けていました。修正しておきます。
カチャカのIPアドレスはわかりますでしょうか？
具体的な例を書いておきます。IPアドレスに応じて適宜書き換えて利用ください。
```
client = kachaka_api.KachakaApiClient("100.94.1.1:26400")
```

**Reply by nozaki-pfr** [reply]
カチャカのIPアドレスは、アプリの 設定 → アプリ情報 → IPアドレス を参照いただくか、
「ねぇカチャカ、IPアドレスを教えて 」と音声指示を出すことで取得可能です

**Reply by hhs0172282066** [reply]
client = kachaka_api.KachakaApiClient("100.94.1.1:26400")

上記コマンド（IPアドレスは有効なものに変更した状態から）を
JupyterLabのターミナルから実行してみましたがSyntax errorになります。
原因はわかりますでしょうか？

**Reply by nozaki-pfr** [reply]
わからないため、エラー内容か再現コードを貼っていただけないでしょうか

**Reply by NorihisaSegawa** [reply]
> カチャカを利用いただきありがとうございます。
> 
> ケースごとにサンプル例を紹介致します。
> 
> 
> 
> ## 外部からカチャカのIPアドレスでカチャカに到達できる場合
> 
> 
> 
> ### gRPCを外部から叩く
> 
> 
> 
> #### kachaka-apiライブラリ
> 
> 
> 
> 例えば外部でpythonプログラムやJuptyerLabを動かして
> 
> ```
> 
> client = kachaka_api.KachakaApiClient("<カチャカのIPアドレス >:26400")
> 
> ```
> 
> とすることでかカチャカの中で動かすのと同様に通信を行うことができます。
> 
> 
> 
> #### gRPC-web
> 
> 
> 
> gRPC-webを利用してwebからkachaka-apiを利用するsampleが下記にあります
> 
> https://github.com/pf-robotics/kachaka-api/tree/main/web/demos/kachaka_api_web_sample
> 
> 
> 
> ### カチャカにサーバを立ててアクセスする
> 
> 
> 
> URLからkachaka-apiを実行する簡単なサンプルが下記にあります。
> 
> https://github.com/pf-robotics/kachaka-api/blob/main/python/demos/url_kachaka_api.ipynb
> 
> 

カチャカにサーバを立ててアクセスする
を試しています。
そのときに使えるコマンド一覧表みたいなのは、存在しますか？（APIとHTTP requestの文字列の関係を探すのがわからなかったので）

**Comment by hhs0172282066** [comment]
READMEにある”Playgroundでサンプルプログラムを実行する”を試しています。

カチャカにsshでログインがうまくいっていないようなのですが、
sshでログインするまでの詳細な手順を教えていただきたいです。

＜やってみたこととして＞
jupyterLabのターミナルからssh-keygenで公開鍵と秘密鍵を生成
↓
公開鍵をコピーしてutils/set_authorized_keys.ipynbへ貼り付け
▶▶ボタンを押す
↓
jupyterLabのターミナルから以下のコマンドに各部入力の上で実行
ssh -p 26500 -i <登録した公開鍵に対応する秘密鍵> kachaka@<kachakaのIPアドレス>
↓
Connection timed outが表示される

utils/set_authorized_keys.ipynbにあるコメント内には　-i　のオプション記載がなく、
-i　無しのコマンドも試してみましたがそちらもタイムアウトになります。

**Reply by nozaki-pfr** [reply]
sshは外部のPCからkachakaにログインするための方法になります。
そのため、kachakaの内部で動いているjupyterLabからのsshは失敗します。
sshした後と同じ状態になっていますので、jupyterLabのターミナルを利用される場合は、カチャカにsshするはスキップして進めることが可能です

**Reply by hhs0172282066** [reply]
外部のPCからkachakaにログインしたいのでsshでログインするまでの詳細な手順を教えていただきたいです。

**Reply by nozaki-pfr** [reply]
外部PCのターミナルからssh-keygenで公開鍵と秘密鍵を生成
↓
KachakaのJupytetrLabに公開鍵をコピーしてutils/set_authorized_keys.ipynbへ貼り付け
▶▶ボタンを押す
↓
外部PCのターミナルから以下のコマンドに各部入力の上で実行
ssh -p 26500 -i <登録した公開鍵に対応する秘密鍵> kachaka@<kachakaのIPアドレス>

のように実行してください。

**Comment by infinitto6-Takira** [comment]
APIを使用し、Pythonを使ってカチャカに動作指示を送るプログラムを作成しています。
"set_robot_stop"というAPIコマンドを見つけたのですが、走行中のカチャカがその場で停止するものだと思っており、目的地に移動中に実行しましたが、止まらず、目的地に到着しました。
一時停止以外でカチャカがその場で停止するコマンドはございますか。

**Reply by inabas-pfr** [reply]
こちらをお試しいただけますでしょうか。
https://github.com/pf-robotics/kachaka-api/blob/main/docs/kachaka_api_client.ipynb
> 実行中のコマンドのキャンセル
> client.cancel_command()

**Reply by infinitto6-Takira** [reply]
ありがとうございます。実行前にコマンド実行中の確認を行い、実行中の場合、上記コマンドを実行したらキャンセルされ、停止コマンド送信で停止を確認できました。
